File: tests/07_phases_ompss.dg/fortran/success_multideps_01.f90
! <testinfo>
! test_generator=config/mercurium-ompss
! test_FFLAGS="--no-copy-deps"
! test_ENV="NX_THROTTLE=dummy"
! test_compile_fail_nanos6_mercurium=yes
! test_compile_fail_nanos6_imfc=yes
! </testinfo>
MODULE MOO
    INTEGER, PARAMETER :: NUM_ITEMS = 100
    INTEGER, PARAMETER :: NUM_TASKS = 10000
    INTEGER VEC(NUM_ITEMS)

    CONTAINS

        SUBROUTINE GENERATE(LENGTH, A, VAL)
            IMPLICIT NONE
            INTEGER :: LENGTH
            INTEGER :: A(LENGTH)
            INTEGER :: VAL

            INTEGER :: I

            !$OMP TASK PRIVATE(I) FIRSTPRIVATE(LENGTH) OUT( [A(K), K = 1, LENGTH] )
            DO I = 1, LENGTH
                A(I) = VAL
            END DO
            !$OMP END TASK

        END SUBROUTINE GENERATE

        SUBROUTINE CONSUME(LENGTH, A, VAL)
            IMPLICIT NONE
            INTEGER :: LENGTH
            INTEGER :: A(LENGTH)
            INTEGER :: VAL

            INTEGER :: I

            !$OMP TASK PRIVATE(I) FIRSTPRIVATE(LENGTH) IN( [A(K), K = 1, LENGTH] )
            DO I = 1, LENGTH
                IF (A(I) /= VAL) STOP 1
            END DO
            !$OMP END TASK

        END SUBROUTINE CONSUME

END MODULE

PROGRAM MAIN
    USE MOO
    IMPLICIT NONE

    INTEGER :: I

    INTEGER :: START, LENGTH, VAL

    PRINT *, "INITIALIZING", NUM_ITEMS, "ITEMS"

    DO I = 1, NUM_ITEMS
        VEC(I) = I
    END DO

    PRINT *, "CREATING", NUM_TASKS, "TASKS"

    DO I = 1, NUM_TASKS
        START = MOD(I, NUM_ITEMS)
        LENGTH = 20

        IF ( START + LENGTH >= NUM_ITEMS ) LENGTH = NUM_ITEMS - START

        VAL = I

        CALL GENERATE(LENGTH, VEC(START), VAL)
        CALL CONSUME(LENGTH, VEC(START), VAL)
    END DO

    PRINT *, "WAITING TASKS"
    !$OMP TASKWAIT
END PROGRAM MAIN

File: tests/07_phases_ompss.dg/cxx/success_multideps_01.cpp
/*--------------------------------------------------------------------
  (C) Copyright 2006-2012 Barcelona Supercomputing Center
                          Centro Nacional de Supercomputacion
  
  This file is part of Mercurium C/C++ source-to-source compiler.
  
  See AUTHORS file in the top level directory for information
  regarding developers and contributors.
  
  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 3 of the License, or (at your option) any later version.
  
  Mercurium C/C++ source-to-source compiler is distributed in the hope
  that it will be useful, but WITHOUT ANY WARRANTY; without even the
  implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
  PURPOSE.  See the GNU Lesser General Public License for more
  details.
  
  You should have received a copy of the GNU Lesser General Public
  License along with Mercurium C/C++ source-to-source compiler; if
  not, write to the Free Software Foundation, Inc., 675 Mass Ave,
  Cambridge, MA 02139, USA.
--------------------------------------------------------------------*/


/*
<testinfo>
test_generator=config/mercurium-ompss
test_ENV="NX_THROTTLE=dummy"
test_CXXFLAGS="--no-copy-deps"
</testinfo>
*/
#include <stdio.h>
#include <assert.h>
#include <unistd.h>
#include <stdlib.h>

void generate0(int length, int *a, int val)
{
#pragma omp task out( { a[k], k = 0;length } )
    {
        usleep(50);
        int i;
        for (i = 0; i < length; i++)
        {
            a[i] = val;
        }
    }
}

void consume0(int length, int *a, int val)
{
#pragma omp task in( { a[k], k = 0;length } )
    {
        usleep(50);
        int i;
        for (i = 0; i < length; i++)
        {
            if (a[i] != val)
            {
                fprintf(stderr, "a[%d] == %d but should be %d\n", i, a[i], val);
                abort();
            }
        }
    }
}

void generate1(int start, int length, int *a, int val)
{
#pragma omp task out( { a[k], k = start;length } )
    {
        usleep(50);
        int i;
        for (i = start; i < (start + length); i++)
        {
            a[i] = val;
        }
    }
}

void consume1(int start, int length, int *a, int val)
{
#pragma omp task in( { a[k], k = start;length } )
    {
        usleep(50);
        int i;
        for (i = start; i < (start + length); i++)
        {
            if (a[i] != val)
            {
                fprintf(stderr, "a[%d] == %d but should be %d\n", i, a[i], val);
                abort();
            }
        }
    }
}


enum { NUM_ITEMS = 100, NUM_TASKS = 10000 };
int vec[NUM_ITEMS];

int main(int argc, char *argv[])
{
    fprintf(stderr, "Initializing %d items...\n", NUM_ITEMS);
    int i;
    for (i = 0; i < NUM_ITEMS; i++)
    {
        vec[i] = i;
    }

    fprintf(stderr, "First round: Creating %d tasks...\n", NUM_TASKS);
    for (i = 0; i < NUM_TASKS; i++)
    {
        int start = (i % NUM_ITEMS);
        int length = 20;
        if ((start + length) >= NUM_ITEMS)
        {
            length = NUM_ITEMS - start;
        }

        int val = i;

        generate0(length, vec + start, val);
        consume0(length, vec + start, val);
    }

    fprintf(stderr, "Second round: Creating %d tasks...\n", NUM_TASKS);
    for (i = 0; i < NUM_TASKS; i++)
    {
        int start = (i % NUM_ITEMS);
        int length = 20;

        if ((start + length) >= NUM_ITEMS)
        {
            length = NUM_ITEMS - start;
        }

        int val = i;

        generate1(start, length, vec, val);
        consume1(start, length, vec, val);
    }
    fprintf(stderr, "Waiting tasks...\n");

#pragma omp taskwait

    return 0;
}
